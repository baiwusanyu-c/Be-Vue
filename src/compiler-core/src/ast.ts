export const enum nodeTypes {
    INTERPOLATION = 'interpolation',
SIMPLE_EXPRESSION = 'simple_expression'
}