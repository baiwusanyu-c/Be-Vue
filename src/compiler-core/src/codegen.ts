import {CREATE_ELEMENT_VNODE, helperMapName, TO_DISPLAY_STRING} from "./runtimeHelpers";
import {nodeTypes} from "./ast";


export function codegen(ast: any) {
    const context = createCodegenContext(ast)
    const {push} = context
    genFunctionPreamble(ast,context);
    push('function ')
    const argList = ['_ctx','_cache']
    const signature = argList.join(', ')
    const functionName = 'render'
    push(`${functionName}( ${signature}){`)
    push('return ')
    genNode(ast.codegenNode,context)
    push(`}`)
    return {
        code:context.code
    }
}
// 生成 文本类型
function genText(node:any,context:any) {
    const {push} = context
    push(node.content)
}

// 生成 差值 {{}}
function genInterpolation(node:any,context:any) {
    const {push,helper} = context
    push(`${helper(helperMapName[TO_DISPLAY_STRING])}`)
    push(`(`)
    genNode(node.content, context)
    push(`)`)
}
// 生成 简单表达式 _ctx.foo
function genSimpleExpression(node: any, context: any) {
    const {push} = context
    push(`${node.content}`)
}

function genElement(node: any, context: any) {
    const {push,helper} = context
    push(`${helper(helperMapName[CREATE_ELEMENT_VNODE])}`)
    push(`(`)
    push(`'${node.tag}'`)
    push(`)`)
}

function genNode(node:any,context:any){
    switch (node.type) {
        case nodeTypes.TEXT:
            genText(node, context);
            break;
        case nodeTypes.INTERPOLATION:
            genInterpolation(node, context);
            break;
        case nodeTypes.SIMPLE_EXPRESSION:
            genSimpleExpression(node, context);
            break;
        case nodeTypes.ELEMENT:
            genElement(node, context);
            break;
        default:
            break;
    }
   
}
// 生成上下文对象
function createCodegenContext(ast:any){
    const context =  {
        code:'',
        push:(source:string)=>{
            context.code += source
        },
        helper:(key:string)=>`_${key}`
    }
    return context
}
// 生成导入语句
function genFunctionPreamble(node:any,context:any) {
    const { push } = context
    const bindging = 'vue'
    const aliasHelpers = ((s:any) => s = `${helperMapName[s as keyof  typeof helperMapName]}: _${helperMapName[s as keyof  typeof helperMapName]}`)
    if(node.helpers.length > 0){
        push(`const { ${node.helpers.map(aliasHelpers).join(', ')} } = ${bindging}`)
    }
    push('\n')
    push('return ')
}