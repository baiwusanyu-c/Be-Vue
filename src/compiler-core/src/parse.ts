// 生成上下文，用于后面的解析
import {nodeTypes} from "./ast";

function createParserContext(content: string) {
    return {
        source:content
    }
}
// 生成ast root
function createRoot(parseChildren: any) {
    return {
        children:parseChildren
    }
}

// 解析 {{}}
function parseInterpolation(context: any) {
    // {{ message }}
    let startDelimiter = '{{'
    let closeDelimiter = '}}'
    // 获取 '}}' 位置
    let closeIndex = context.source.indexOf(closeDelimiter,startDelimiter.length)
    // 消费 '{{'
    advanceBy(context,startDelimiter.length)
    // 截取 content
    let rawContent = context.source.slice(0,closeIndex - startDelimiter.length)
    let content = rawContent.trim()
    // 消费 '}}'
    advanceBy(context,rawContent.length + closeDelimiter.length)
    return {
        type:nodeTypes.INTERPOLATION,
        content:{
            type:nodeTypes.SIMPLE_EXPRESSION,
            content:content
        }
    }
}
function advanceBy(context:any,index:any) {
    context.source = context.source.slice(index)
}
// 解析 children
function parseChildren(context: any) {
    const nodes = []
    let node
    // 以 {{ 则走解析插值逻辑
    if(context.source.startsWith('{{')){
        node = parseInterpolation(context)
    }
    nodes.push(node)
    return nodes
}


export function baseParse(content:string) {
    const context = createParserContext(content)
    return createRoot(parseChildren(context))
}