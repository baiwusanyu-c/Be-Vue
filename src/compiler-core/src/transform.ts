import {nodeTypes} from "./ast";
import {TO_DISPLAY_STRING} from "./runtimeHelpers";

function createRootCodegen(root:any) {
    root.codegenNode = root.children[0]
}

export function transform(root:any,option:any = {}){
    const context = createTransFormContext(root,option)
    traverseNode(root,context)
    createRootCodegen(root)
    root.helpers = [...context.helpers.keys()]
}

function traverseChildren(root: any, context: any) {
    const children = root.children
    for (let i = 0; i < children.length; i++) {
        traverseNode(children[i], context)
    }

}
// 遍历ast，并调用 转换函数
export function traverseNode(root:any,context:any){
    // 执行传入的转换函数
    const nodeTransforms = context.nodeTransforms
    if(nodeTransforms){
        for (let i = 0;i < nodeTransforms.length;i++){
            nodeTransforms[i](root,context)
        }
    }
    switch (root.type) {
        case nodeTypes.INTERPOLATION:
            context.helper(TO_DISPLAY_STRING)
            break;
        case nodeTypes.ELEMENT:
        case nodeTypes.ROOT:
            // 深度优先遍历ast
            traverseChildren(root, context);
            break;
        default:
            break;
    }


}
// 生成 transform 上下文
export function createTransFormContext(root:any,option:any){
    const context =  {
        root,
        nodeTransforms:option.nodeTransforms || [],
        helpers:new Map(),
        helper:(key:string)=>{
            context.helpers.set(key,1)
        }
    }
    return context
}