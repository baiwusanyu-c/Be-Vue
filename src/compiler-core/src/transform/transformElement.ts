import {nodeTypes} from "../ast";
import {CREATE_ELEMENT_VNODE} from "../runtimeHelpers";

export function transformElement(node:any,context:any){
    if (node.type === nodeTypes.ELEMENT){
        context.helper(CREATE_ELEMENT_VNODE)
    }
}