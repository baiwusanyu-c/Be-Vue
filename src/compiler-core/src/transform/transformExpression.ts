import {nodeTypes} from "../ast";

export function transformExpression(node:any){
    if (node.type === nodeTypes.ELEMENT){
        node.content = processExpression(node.content)
    }
}
function processExpression(node:any){
     node.content = `_ctx.${node.content}`
     return node
}