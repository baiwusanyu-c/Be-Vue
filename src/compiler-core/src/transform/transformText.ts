import {nodeTypes} from "../ast";
import {isText} from "../utils";

export function transformText(node: any, context: any) {
    if (node.type === nodeTypes.ELEMENT) {
        return () => {
            const {children} = node
            let currentContainer
            // 遍历节点 children，把所有子节点全部收集到 COMPOUND_EXPRESSION 类型节点中（还 添加了 + ）
            // 收集完成后 并把 COMPOUND_EXPRESSION 类型节点 替换成 children
            for (let i = 0; i < children.length; i++) {
                let child = children[i]
                if (isText(child)) {
                    for (let j = i + 1; j < children.length; j++) {
                        const next = children[j]
                        if (isText(next)) {
                            if (!currentContainer) {
                                currentContainer = children[i] = {
                                    type: nodeTypes.COMPOUND_EXPRESSION,
                                    children: [child]
                                }
                            }
                            currentContainer.children.push(' + ')
                            currentContainer.children.push(next)
                            children.splice(j, 1)
                            j--

                        } else {
                            currentContainer = undefined
                            break;
                        }
                    }
                }
            }
        }
    }
}