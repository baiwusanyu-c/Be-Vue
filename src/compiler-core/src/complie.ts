import {baseParse} from "./parse";
import {transform} from "./transform";
import {transformExpression} from "./transform/transformExpression";
import {transformElement} from "./transform/transformElement";
import {transformText} from "./transform/transformText";
import {generate} from "./codegen";

export function baseCompile(template:string){
    const ast = baseParse(template)
    transform(ast,{
        nodeTransforms:[transformExpression,transformElement,transformText]
    })
    return  generate(ast)
}