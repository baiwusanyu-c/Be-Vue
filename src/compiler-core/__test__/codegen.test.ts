
import {baseParse} from '../src/parse'
import {codegen} from '../src/codegen'
import {transform} from "../src/transform";
import {transformExpression} from "../src/transform/transformExpression";
import {transformElement} from "../src/transform/transformElement";
describe('codegen',()=>{
    test('string',()=>{
        const ast = baseParse('czh')
        transform(ast)
        const node = codegen(ast)
        // 快照测试
        expect(node).toMatchSnapshot()
    })
    test('interpolation',()=>{
        const ast = baseParse('{{czh}}')
        transform(ast,{
            nodeTransforms:[transformExpression]
        })
        const node = codegen(ast)
        // 快照测试
        expect(node).toMatchSnapshot()
    })
    test('element',()=>{
        const ast = baseParse('<div></div>')
        transform(ast,{
            nodeTransforms:[transformElement]
        })
        const node = codegen(ast)
        // 快照测试
        expect(node).toMatchSnapshot()
    })

})