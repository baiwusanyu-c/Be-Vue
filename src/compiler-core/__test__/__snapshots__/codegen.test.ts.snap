// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`codegen element 1`] = `
Object {
  "code": "const { createElementVNode: _createElementVNode } = vue
return function render( _ctx, _cache){return _createElementVNode('div')}",
}
`;

exports[`codegen interpolation 1`] = `
Object {
  "code": "const { toDisplayString: _toDisplayString } = vue
return function render( _ctx, _cache){return _toDisplayString(_ctx.czh)}",
}
`;

exports[`codegen string 1`] = `
Object {
  "code": "
return function render( _ctx, _cache){return czh}",
}
`;
